File: app/src/main/java/net/feheren_fekete/applist/applistpage/IconCache.java
package net.feheren_fekete.applist.applistpage;

import android.graphics.Bitmap;

import net.feheren_fekete.applist.applistpage.viewmodel.AppItem;

import androidx.collection.LruCache;

public class IconCache {
    private LruCache<String, Bitmap> mCache;

    public IconCache() {
        // Get max available VM memory, exceeding this amount will throw an
        // OutOfMemory exception. Stored in kilobytes as LruCache takes an
        // int in its constructor.
        final int maxMemory = (int) (Runtime.getRuntime().maxMemory() / 1024);

        // Use 1/8th of the available memory for this memory cache.
        final int cacheSize = maxMemory / 8;

        mCache = new LruCache<String, Bitmap>(cacheSize) {
            @Override
            protected int sizeOf(String key, Bitmap bitmap) {
                // The cache size will be measured in kilobytes rather than
                // number of items.
                return bitmap.getByteCount() / 1024;
            }
        };
    }

    public void addIcon(String key, Bitmap bitmap) {
        if (getIcon(key) == null) {
            mCache.put(key, bitmap);
        }
    }

    public Bitmap getIcon(String key) {
        return mCache.get(key);
    }

    public String createKey(AppItem appItem) {
        return appItem.getPackageName() + "::" + appItem.getClassName();
    }
}

File: app/src/main/java/net/feheren_fekete/applist/applistpage/IconCache.kt
package net.feheren_fekete.applist.applistpage

import android.graphics.Bitmap

import net.feheren_fekete.applist.applistpage.viewmodel.AppItem

import androidx.collection.LruCache

class IconCache {
    private val cache: LruCache<String, Bitmap>

    init {
        // Get max available VM memory, exceeding this amount will throw an
        // OutOfMemory exception. Stored in kilobytes as LruCache takes an
        // int in its constructor.
        val maxMemory = (Runtime.getRuntime().maxMemory() / 1024).toInt()

        // Use 1/8th of the available memory for this memory cache.
        val cacheSize = maxMemory / 8

        cache = object : LruCache<String, Bitmap>(cacheSize) {
            override fun sizeOf(key: String, bitmap: Bitmap): Int {
                // The cache size will be measured in kilobytes rather than
                // number of items.
                return bitmap.byteCount / 1024
            }
        }
    }

    fun addIcon(key: String, bitmap: Bitmap) {
        if (getIcon(key) == null) {
            cache.put(key, bitmap)
        }
    }

    fun getIcon(key: String): Bitmap? {
        return cache.get(key)
    }

    fun createKey(appItem: AppItem): String {
        return appItem.packageName + "::" + appItem.className
    }
}
